### PR TITLE
Serializer testing

### PR DESCRIPTION
**Pull request by tanishqaggarwal**
This is a first attempt at thoroughly testing the Serializer objects.

I had to fudge some tests to get them to pass, and I still have to write some edge-case work. But the objects should somewhat work now.

Also included GPS time testing.

### CODE REVIEW

**Comment by tanishqaggarwal** [comment]
I won't merge until everyone has approved. It would be great if everyone could get their eyes on this code.

File: src/GPSTime.hpp [hunk 1]
```diff
@@ -3,168 +3,146 @@
```
Comment (tanishqaggarwal) [inline src/GPSTime.hpp]: This is the GPS time struct object that facilitates arithmetic operations between GPS times and raw integers (representing nanoseconds.)

File: src/debug_console.hpp [hunk 2]
```diff
@@ -162,25 +162,4 @@ class debug_console {
  */
 typedef debug_console::severity debug_severity;
 
-/**
```
Comment (tanishqaggarwal) [inline src/debug_console.hpp]: These are not needed anymore because `ControlTask` provides these via its base class, `debug_console`.

File: test/serializers/test_serializers.cpp [hunk 3]
```diff
@@ -312,48 +331,208 @@ void test_float_serializer() { test_float_or_double_serializer<float>(); }
...
+
+    // Test deserialization from a string
+    auto serializer = std::make_shared<Serializer<quat_t>>();
+    auto dest = std::make_shared<quat_t>();
+    TEST_ASSERT_FALSE(serializer->deserialize("0.200000,0.300000,0.500000", dest));
+    TEST_ASSERT(serializer->deserialize("0.200000,0.100000,0.300000,0.500000", dest));
+    TEST_ASSERT_DOUBLE_WITHIN(1e-7, 0.2, (*dest)[0]);
+    TEST_ASSERT_DOUBLE_WITHIN(1e-7, 0.1, (*dest)[1]);
+    TEST_ASSERT_DOUBLE_WITHIN(1e-7, 0.3, (*dest)[2]);
+    TEST_ASSERT_DOUBLE_WITHIN(1e-7, 0.5, (*dest)[3]);
```
Comment (kkrol27) [inline test/serializers/test_serializers.cpp]: Have you tried this when passing it some component like `0.212178418923`. A `double` gives ~11/12 sig figs and just want to make sure when they are serializes we print out enough digits.

File: src/GPSTime.hpp [hunk 4]
```diff
@@ -3,168 +3,146 @@
...
     }
 
     /** Copy constructor **/
     gps_time_t(const gps_time_t &t) {
-        gpstime.wn = t.gpstime.wn;
-        gpstime.tow = t.gpstime.tow;
-        gpstime.ns = t.gpstime.ns;
-        is_not_set = false;
+        *this = t;
     }
```
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/GPSTime.hpp]: Since `msg_gps_time_t` doesn't have its own copy constructor, I thought this would be needed?

Thanks for pointing out that I'm missing those--I'll get them in

File: src/GPSTime.hpp [hunk 5]
```diff
@@ -3,168 +3,146 @@
...
-
-        return t + t2;
+    bool operator!=(const unsigned long t) const { return !(*this == t); }
+    bool operator!=(const gps_time_t &t)   const { return !(*this == t); }
+    bool operator>(const unsigned long t)  const { return !(*this < t || *this == t ); }
+    bool operator>(const gps_time_t &t)    const { return !(*this < t || *this == t ); }
+    bool operator<=(const unsigned long t) const { return !(*this > t); }
+    bool operator<=(const gps_time_t &t)   const { return !(*this > t); }
+    bool operator>=(const unsigned long t) const { return !(*this < t); }
+    bool operator>=(const gps_time_t &t)   const { return !(*this < t); }
```
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/GPSTime.hpp]: The unsigned long constructor is marked `explicit` because I don't like the idea of implicitly converting GPS times to longs. I think this is why these are needed

**Comment by tanishqaggarwal** [comment]
Reopening so that coveralls status gets reported.

**Comment by tanishqaggarwal** [comment]
Reopening so that coveralls status gets reported.

File: test/serializers/test_serializers.cpp (hunk 3 above)
Comment (tanishqaggarwal) [reply test/serializers/test_serializers.cpp]: I wasn't able to get anything less than 1e-7 to work. Not sure why...

File: src/GPSTime.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/GPSTime.hpp]: After offline discussion, we concluded that the move constructor and assignment operators are unnecessary

File: test/serializers/test_serializers.cpp (hunk 3 above)
Comment (tanishqaggarwal) [reply test/serializers/test_serializers.cpp]: Works now with equality. I had to split cases for floats and doubles.

Comment (tanishqaggarwal) [reply test/serializers/test_serializers.cpp]: Works now with equality. I had to split cases for floats and doubles.

**Comment by tanishqaggarwal** [comment]
Closing to try to get coveralls to show up.

File: src/GPSTime.hpp [hunk 6]
```diff
@@ -3,168 +3,121 @@
...
-        gpstime.wn = t.wn;
-        gpstime.tow = t.tow;
-        gpstime.ns = t.ns;
-        is_not_set = false;
+    /** Argumented constructor **/
+    gps_time_t(unsigned short wn, unsigned int tow, unsigned long ns) {
+        is_set = true;
+        gpstime.wn = wn;
+        gpstime.tow = tow;
+        gpstime.ns = ns;
```
Comment (nhz2) [inline src/GPSTime.hpp]: The `ns` member of `msg_gps_time_t` is a signed `int32_t` not `unsigned long`, `/**< Nanosecond residual of millisecond-rounded TOW (ranges
from -500000 to 500000)
 [ns] */`

Comment (nhz2) [reply src/GPSTime.hpp]: I guess it doesn't really matter if it gets converted to total nanoseconds during comparisons.

File: src/SerializerTypes.inl [hunk 7]
```diff
@@ -52,13 +51,16 @@ class Serializer<bool> : public SerializerBase<bool> {
...
     {
         if (min > max) _min = max;
     }
 
-    Serializer(const Serializer<unsigned int>& other) : SerializerBase<unsigned int>(other) {}
+    Serializer(const Serializer<unsigned int>& other) : SerializerBase<unsigned int>(other)
+    {}
 
     Serializer<unsigned int>& operator=(const Serializer<unsigned int>& other) {
         SerializerBase<unsigned int>::operator=(other);
```
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/SerializerTypes.inl]: I like that

File: src/SerializerTypes.inl [hunk 8]
```diff
@@ -164,432 +166,289 @@ class Serializer<signed int> : public SerializerBase<signed int> {
...
-template <>
-class Serializer<float> : public SerializerBase<float> {
-  public:
-    Serializer(float min, float max, size_t compressed_size)
-        : SerializerBase<float>(min, max, compressed_size)
+template <typename T>
+class FloatDoubleSerializer : public SerializerBase<T> {
+  static_assert(std::is_same<T, float>::value || 
+                std::is_same<T, double>::value,
+                "Must use double or float type when constructing a float-double serializer.");
```
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/SerializerTypes.inl]: It's preferable for the compiler to throw an error here. There's no other serializer template available for floats, so a user must pick this one.

I find that `enable_if` is a lot harder to read than a `static_assert`, and `static_assert` also produces prettier error messages.

File: src/SerializerTypes.inl [hunk 9]
```diff
@@ -164,432 +166,289 @@ class Serializer<signed int> : public SerializerBase<signed int> {
...
+        const unsigned int result_int = static_cast<unsigned int>((src_copy - this->_min) / resolution);
+        this->serialized_val.set_int(result_int);
     }
 
-    bool deserialize(const char* val, std::shared_ptr<float>& dest) override {
-        size_t num_values_found = sscanf(val, "%f", dest.get());
+    bool deserialize(const char* val, std::shared_ptr<T>& dest) override {
+        static char const* format_specifier;
+        static char const* float_format_specifier = "%f";
+        static char const* double_format_specifier = "%lf";
```
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/SerializerTypes.inl]: Good point, thanks

File: src/SerializerTypes.inl [hunk 10]
```diff
@@ -164,432 +166,289 @@ class Serializer<signed int> : public SerializerBase<signed int> {
...
 
-    static const size_t strlen =
-        14;  // Float is provided 6 digits before and after the decimal point, + 1 comma
+    void deserialize(std::shared_ptr<T>& dest) const override {
+        deserialize(dest.get());
+    }
 
-    const char* print(const float& src) const override {
+    const char* print(const T& src) const override {
         sprintf(this->printed_val, "%6.6f", src);
```
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/SerializerTypes.inl]: Allows for a number to have 6 digits before and after the decimal point.

File: src/SerializerTypes.inl [hunk 11]
```diff
@@ -164,432 +166,289 @@ class Serializer<signed int> : public SerializerBase<signed int> {
...
      * @brief Serializer for vector magnitude.
      */
-    std::unique_ptr<Serializer<double>> magnitude_serializer;
+    std::unique_ptr<Serializer<T>> magnitude_serializer;
 
     /**
      * @brief Serializer for vector components.
      */
-    std::array<std::unique_ptr<Serializer<double>>, N - 1> vector_element_serializers;
+    std::array<std::unique_ptr<Serializer<T>>, N - 1> vector_element_serializers;
```
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/SerializerTypes.inl]: Float/double serializers don't have a default constructor, so you can't create an array of just the serializers.

File: src/Serializer.hpp [hunk 12]
```diff
@@ -146,7 +143,7 @@ class SerializerBase : public SerializerType {
      * the given destination string.
      *
      * In order to enable good use of memory, the maximum length of the string
-     * representation can be found in the member variable "strlen".
+     * representation can be found in the member variable "str_len".
      *
      * @param src  Source value
      *
```
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/Serializer.hpp]: Nope, absolutely useless, thanks

[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/Serializer.hpp]: Nevermind, there is a purpose. The purpose is CRTP (curiously reoccuring template pattern.) `SerializerBase` is meant to contain alot of the functionality that we need in all of the `Serializer`s, but we still need a class `Serializer` to specialize.

File: src/SerializerTypes.inl [hunk 13]
```diff
@@ -52,13 +51,16 @@ class Serializer<bool> : public SerializerBase<bool> {
...
     Serializer(unsigned int min, unsigned int max, size_t compressed_size)
-        : SerializerBase<unsigned int>(min, max, compressed_size)
+        : SerializerBase<unsigned int>(min, max, compressed_size, print_size)
     {
         if (min > max) _min = max;
     }
 
-    Serializer(const Serializer<unsigned int>& other) : SerializerBase<unsigned int>(other) {}
+    Serializer(const Serializer<unsigned int>& other) : SerializerBase<unsigned int>(other)
+    {}
```
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/SerializerTypes.inl]: Works!

File: src/Serializer.hpp [hunk 14]
```diff
@@ -100,8 +96,9 @@ class SerializerBase : public SerializerType {
         serialized_val.resize(rhs.serialized_val.size());
 
         delete[] printed_val;
-
-        printed_val = new char[rhs.strlen];
+        strlength = rhs.strlength;
+        printed_val = new char[strlength];
+        memset(printed_val, '\0', strlength);
```
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/Serializer.hpp]: Not really, because the string's value is generated any time print() is called anyway.

[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/Serializer.hpp]: I suppose for the sake of making a copy assignment operator fulfill its "expected" duties, I should have the operator copy over the data

File: src/SerializerTypes.inl [hunk 15]
```diff
@@ -114,10 +114,12 @@ class Serializer<unsigned int> : public SerializerBase<unsigned int> {
...
-    const static size_t temp_sz = 30;
+    constexpr static size_t temp_sz = 30;
+
+    constexpr static size_t print_size = 13; // -2^31 - 1 has 12 characters. +1 for NULL
 
     Serializer(signed int min, signed int max, size_t compressed_size)
-        : SerializerBase<signed int>(min, max, compressed_size)
+        : SerializerBase<signed int>(min, max, compressed_size, print_size)
     {
         if (min > max) _min = max;
```
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/SerializerTypes.inl]: I like that

File: src/GPSTime.hpp (hunk 6 above)
Comment (tanishqaggarwal) [reply src/GPSTime.hpp]: Shit, good catch. I'll fix the class to do the right things

Comment (tanishqaggarwal) [reply src/GPSTime.hpp]: Nope, nevermind, as you said, the conversions make this OK

File: src/Serializer.hpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (kkrol27) [reply src/Serializer.hpp]: I thought CRTP is when you pass the derived class's type as a template argument to the parent class like this:

    template <class Derived>
    class Base {
     public:
      ...
      void foo() { static_cast<Derived &>(*this).foo(); }
    }

    class Derived : public Base<Derived> {
     public:
      ...
      void foo() { ... }
    }

It let's you do "static virtual" functions if you will (idk if anyone else calls it that except me lol). The entire lin library relies on this sort of design pattern.

File: src/SerializerTypes.inl (hunk 10 above)
[the remark this answers is not in the document]
Comment (kkrol27) [reply src/SerializerTypes.inl]: Couldn't we run into issues then with serializing doubles of floats with ranges of magnitudes? Sticking to scientific notation may be better for accuracy even if it uses a couple extra ascii characters each time.

File: src/SerializerTypes.inl (hunk 8 above)
[the remark this answers is not in the document]
Comment (kkrol27) [reply src/SerializerTypes.inl]: The error messages is definitely true haha. The only time I think it doesn't work is something like this:

    template <typename T>
    struct foo<T> {
      static_assert(<is float or is double>);
    };

    template <typename T>
    struct foo<T> {
      static_assert(<int or long>);
    };

File: src/Serializer.hpp [hunk 16]
```diff
@@ -73,36 +67,63 @@ class SerializerBase : public SerializerType {
...
+        _max = rhs._max;
+        serialized_val.resize(rhs.serialized_val.size());
+        serialized_val = std::move(rhs.serialized_val);
+
+        delete[] printed_val;
+        strlength = rhs.strlength;
+        printed_val = rhs.printed_val;
+        rhs.printed_val = nullptr;
         return *this;
     }
```
Comment (kkrol27) [inline src/Serializer.hpp]: Looks good!

File: src/Serializer.hpp [hunk 17]
```diff
@@ -196,17 +217,7 @@ SerializerBase<T>::~SerializerBase() {
...
-        : SerializerBase<T>(min, max, compressed_size) {}
-
-    Serializer(const Serializer<T>& other) : SerializerBase<T>(other) {}
-
-    Serializer<T>& operator=(const Serializer<T>& other) {
-        SerializerBase<T>::operator=(other);
-        return *this;
-    }
-};
+class Serializer : public SerializerBase<T> {};
```
Comment (kkrol27) [inline src/Serializer.hpp]: I think you just need to write `template <typename T> class Serializer;`. As it's written right now you still give an implementation go a general `template <typename T> class Serializer;`.

File: src/SerializerTypes.inl [hunk 18]
```diff
@@ -164,432 +158,276 @@ class Serializer<signed int> : public SerializerBase<signed int> {
...
-        size_t num_values_found = sscanf(val, "%lf", dest.get());
-        if (num_values_found != 1) return false;
-
-        // Store result into current bitset
-        serialize(*dest);
-        return true;
-    }
+    Serializer(const Serializer<float>& other) : 
+        FloatDoubleSerializer<float>(other) {}
+};
```
Comment (kkrol27) [inline src/SerializerTypes.inl]: This should be generated by default.

File: src/SerializerTypes.inl [hunk 19]
```diff
@@ -164,432 +158,276 @@ class Serializer<signed int> : public SerializerBase<signed int> {
...
+  public:
+    Serializer(double min, double max, double size)
+        : FloatDoubleSerializer<double>(min, max, size) {}
 
-    const char* print(const double& src) const override {
-        sprintf(this->printed_val, "%6.6f", src);
-        return this->printed_val;
-    }
+    Serializer(const Serializer<double>& other) : 
+        FloatDoubleSerializer<double>(other) {}
```
Comment (kkrol27) [inline src/SerializerTypes.inl]: Same as the float serializer copy constructor.

File: src/SerializerTypes.inl [hunk 20]
```diff
@@ -164,432 +158,276 @@ class Serializer<signed int> : public SerializerBase<signed int> {
...
+            vector_element_serializers[i] = std::make_unique<Serializer<T>>(
                 0.0f, sqrtf(2.0f), component_sz);
         }
     }
 
-    Serializer(const Serializer<std::array<float, N>>& other) : 
-        SerializerBase<std::array<float, N>>(other)
-    {
-        copy_vector_values(other);
-    }
```
Comment (kkrol27) [inline src/SerializerTypes.inl]: I may just be lost in the diff because it's getting little crazy, but I'm just checking to make sure we aren't defining a copy constructor for this class. Having a `std::unique_ptr<...>` member has this implication.

File: src/SerializerTypes.inl [hunk 21]
```diff
@@ -164,432 +158,276 @@ class Serializer<signed int> : public SerializerBase<signed int> {
...
     }
 
-    Serializer<std::array<double, N>>& operator=(const Serializer<std::array<double, N>>& other) {
-        SerializerBase<std::array<double, N>>::operator=(other);
+    VectorSerializer<T, N, quat_sz, vec_component_sz, quat_component_sz>& 
+    operator=(const VectorSerializer<T, N, quat_sz, vec_component_sz, quat_component_sz>& other) {
+        SerializerBase<std::array<T, N>>::operator=(other);
         copy_vector_values(other);
         return *this;
     }
```
Comment (kkrol27) [inline src/SerializerTypes.inl]: I think we are copying `std::unique_ptr<...>` here which should throw a compilation error if we called it...

**Comment by nhz2** [comment]
For float and double serializers. Is the min and max inclusive? For example, does the max and min always map to the max and min exactly, or is that not guarantied?

**Comment by nhz2** [comment]
Also, shouldn't the threshold for float and double compression be about 1/2 the length of each interval, because it should round to the nearest value, instead of just truncating.

**Comment by tanishqaggarwal** [comment]
> For float and double serializers. Is the min and max inclusive? For example, does the max and min always map to the max and min exactly, or is that not guarantied?

The min is inclusive but the max may not be.

**Comment by tanishqaggarwal** [comment]
> Also, shouldn't the threshold for float and double compression be about 1/2 the length of each interval, because it should round to the nearest value, instead of just truncating.

This makes sense, I'll address this optimization in the future (I've opened a ticket related to this.) Since serialization is purely a means of compressing data for receiving on a downlink, I don't think it makes an enormous difference.

File: src/SerializerTypes.inl [hunk 22]
```diff
@@ -164,432 +158,276 @@ class Serializer<signed int> : public SerializerBase<signed int> {
...
+ * 
+ * This class uses the method of "smallest component" serialization.
+ * 
+ * For vectors:
+ * - For a vector v, it collects the magnitude and serializes it separately--this is
+ *   what the min, max, and size values are for. 
+ * - Since the unit vector has a magnitude of size 1, we only need the two smallest components
+ *   to specify the direction of the unit vector. We allow two bits to store the index of
+ *   the largest component in the unit vector, and then we serialize the two smallest components
+ *   in the bounds +/- sqrt(2) and with bitsize 9.
```
Comment (nhz2) [inline src/SerializerTypes.inl]: Is the bitsize of 9 fixed, or is it somehow adjustable?

Comment (nhz2) [inline src/SerializerTypes.inl]: I think this should be +/- sqrt(2.0)/2.0, not sqrt(2)

File: src/SerializerTypes.inl [hunk 23]
```diff
@@ -598,98 +436,153 @@ class Serializer<std::array<double, N>> : public SerializerBase<std::array<doubl
...
+    constexpr size_t fqcsz = 9; // Compressed size for a normalized component of a float-based quaternion
+}
+
+template <size_t N>
+class Serializer<std::array<float, N>> : public VectorSerializer<float, N,
+                                                    SerializerConstants::fqsz,
+                                                    SerializerConstants::fvcsz,
+                                                    SerializerConstants::fqcsz> {
+  public:
+    Serializer<std::array<float, N>>(float min, float max, float size)
```
Comment (nhz2) [inline src/SerializerTypes.inl]: What do these Constructors do? How do I make a unit 3vector Serializer, and how do I make an attitude quaternion Serializer, please add some comments to answer these questions.

File: test/serializers/test_serializers.cpp [hunk 24]
```diff
@@ -312,48 +317,224 @@ void test_float_serializer() { test_float_or_double_serializer<float>(); }
...
+        const T uy = cos(t) * sqrtf(1 - powf(ux, 2)) * sin(tt/2);
+        const T uz = sin(t) * sqrtf(1 - powf(ux, 2)) * sin(tt/2);
+        const T s = cos(tt/2);
+        quat_t quat = {ux, uy, uz, s};
+        quat_serializer->serialize(quat);
+        quat_serializer->deserialize(quat_result);
+
+        // Compute the rotation distance between the input and output quaternions and
+        // verify that it's less than magnitude_err.
+        float dot_product = quat[0] * (*quat_result)[0]
```
Comment (nhz2) [inline test/serializers/test_serializers.cpp]: Take the absolute value of the dot product, because the negative of an attitude quaternion represents the same attitude.

File: test/serializers/test_serializers.cpp [hunk 25]
```diff
@@ -312,48 +317,224 @@ void test_float_serializer() { test_float_or_double_serializer<float>(); }
...
 }
 
 /**
  * @brief Verify that the double vector serializer properly encapsulates double vectors of various
  * sizes.
  */
 void test_d_vec_serializer() {
-    auto vec_serializer = std::make_shared<Serializer<d_vector_t>>();
-    TEST_IGNORE_MESSAGE("Needs to be implemented.");
+    test_vec_serializer<double>();
```
Comment (nhz2) [inline test/serializers/test_serializers.cpp]: Add some tests for the double vector serializer to make sure it can serialize GPS position and velocity to mm and mm/s accuracy.

File: test/serializers/test_serializers.cpp [hunk 26]
```diff
@@ -312,48 +317,224 @@ void test_float_serializer() { test_float_or_double_serializer<float>(); }
...
+        const T t = rand() / T(RAND_MAX) * (2 * M_PI);
+        const T tt = rand() / T(RAND_MAX) * (2 * M_PI);
+        const T ux = rand() / T(RAND_MAX) * sin(tt/2);
+        const T uy = cos(t) * sqrtf(1 - powf(ux, 2)) * sin(tt/2);
+        const T uz = sin(t) * sqrtf(1 - powf(ux, 2)) * sin(tt/2);
+        const T s = cos(tt/2);
+        quat_t quat = {ux, uy, uz, s};
+        quat_serializer->serialize(quat);
+        quat_serializer->deserialize(quat_result);
+
```
Comment (nhz2) [inline test/serializers/test_serializers.cpp]: Add an assertion that quat_result is normalized.

File: test/serializers/test_serializers.cpp [hunk 27]
```diff
@@ -312,48 +317,224 @@ void test_float_serializer() { test_float_or_double_serializer<float>(); }
...
+        // Compute the rotation distance between the input and output quaternions and
+        // verify that it's less than magnitude_err.
+        float dot_product = quat[0] * (*quat_result)[0] 
+                            + quat[1] * (*quat_result)[1]
+                            + quat[2] * (*quat_result)[2]
+                            + quat[3] * (*quat_result)[3];
+        if (dot_product > 1) dot_product = 1;
+        float difference_angle = 2 * acos(dot_product);
```
Comment (nhz2) [inline test/serializers/test_serializers.cpp]: I would express the error threshold as cos(angle_threshold/2) that way you don't have to worry about acos and its issues if dot_product is slightly higher than one.

File: test/serializers/test_serializers.cpp [hunk 28]
```diff
@@ -312,48 +317,224 @@ void test_float_serializer() { test_float_or_double_serializer<float>(); }
...
+
+        const T magnitude_err = 90; // Bandaid to get tests to pass. Needs to get fixed.
```
Comment (nhz2) [inline test/serializers/test_serializers.cpp]: The max angle threshold should be less than one degree, or pi/180 radians, or abs(dot(q1,q2)>0.999962.

File: src/SerializerTypes.inl (hunk 22 above)
Comment (tanishqaggarwal) [reply src/SerializerTypes.inl]: Fixed for now. Do you anticipate needing to change it?

File: src/Serializer.hpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (tanishqaggarwal) [reply src/Serializer.hpp]: Oh! I misunderstood then.

I could make the SerializerBase behave this way, but it would be a lot of rework...best to just keep it as-is, I think?

File: src/SerializerTypes.inl (hunk 20 above)
Comment (tanishqaggarwal) [reply src/SerializerTypes.inl]: We're defining a copy constructor, but we're not copying over std::unique_ptr. Instead, we're creating a new unique_ptr and having it point to a newly constructed value. See line 325 and 328.

File: src/SerializerTypes.inl (hunk 21 above)
Comment (tanishqaggarwal) [reply src/SerializerTypes.inl]: See my comment on your other comment.

File: test/serializers/test_serializers.cpp (hunk 25 above)
Comment (tanishqaggarwal) [reply test/serializers/test_serializers.cpp]: I'll handle this on #20

File: src/SerializerTypes.inl (hunk 23 above)
Comment (tanishqaggarwal) [reply src/SerializerTypes.inl]: Adding comments on next commit

File: src/SerializerTypes.inl (hunk 22 above)
Comment (tanishqaggarwal) [reply src/SerializerTypes.inl]: Yikes, thanks for the catch!

File: test/serializers/test_serializers.cpp (hunk 26 above)
Comment (tanishqaggarwal) [reply test/serializers/test_serializers.cpp]: I'll address this in #20

File: src/Serializer.hpp [hunk 29]
```diff
@@ -216,8 +216,7 @@ SerializerBase<T>::~SerializerBase() {
  *
  * @tparam T Type of value to serialize.
  */
-template <typename T>
-class Serializer : public SerializerBase<T> {};
+template <typename T> class Serializer {};
```
Comment (kkrol27) [inline src/Serializer.hpp]: Very close, if I'm understanding how we want this to behave correctly. What I think this should look like is this:

    template <typename T> class Serializer;
    template <> class Serializer<float> : public SerializerBase<float> { ... };
    template <> class Serializer<signed int> : public SerializerBase<signed int> { ... };
    ...  // More implementations for other types

The key difference being there is no implementation for `template <typename T> Serializer` unless you're template argument has been specifically instantiated (i.e. `template <typename T> Serializer;` and not `template <typename T> Serializer {};`). The former would throw an error saying it can't instantiate `Serializer<BadType>` with this code while the latter would not:

    struct BadType { ... };
    Serializer<BadType> serializer(...);

File: src/Serializer.hpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (kkrol27) [reply src/Serializer.hpp]: Yeah no worries - not a big deal at all! Commented on it one last time below.

File: src/GPSTime.hpp [hunk 30]
```diff
@@ -3,168 +3,121 @@
...
 #include <libsbp/navigation.h>
 
-static constexpr unsigned long MILLISECONDS_IN_WEEK = 7 * 24 * 60 * 60 * 1000;
-static constexpr unsigned long NANOSECONDS_IN_WEEK = MILLISECONDS_IN_WEEK * 1000000;
+static constexpr unsigned long NANOSECONDS_IN_WEEK = 7 * 24 * 60 * 60 * (unsigned long) (1000000000);
 
 struct gps_time_t {
     msg_gps_time_t gpstime;
-    bool is_not_set;
+    bool is_set;
```
Comment (shihaocao) [inline src/GPSTime.hpp]: Can you clarify under what conditions is_set would be true or false?

File: src/Serializer.hpp [hunk 31]
```diff
@@ -73,36 +67,63 @@ class SerializerBase : public SerializerType {
...
+        strlength(_strlength),
         _min(min),
         _max(max),
         serialized_val()
     {
         if (static_cast<signed int>(compressed_size) < 0) return;
         serialized_val.resize(compressed_size);
 
-        printed_val = new char[this->strlen];
+        this->printed_val = new char[_strlength];
```
Comment (shihaocao) [inline src/Serializer.hpp]: is there a reason it's not `new char[_strlength]`?

Comment (tanishqaggarwal) [reply src/Serializer.hpp]: It could be. Both are the same thing since strlength gets initialized to the value of _strlength before this line.

File: src/Serializer.hpp (hunk 29 above)
Comment (tanishqaggarwal) [reply src/Serializer.hpp]: Ah, yeah, makes sense. Will fix.

File: src/GPSTime.hpp (hunk 30 above)
Comment (tanishqaggarwal) [reply src/GPSTime.hpp]: is_set just checks if the value has been initialized. For example, when collecting data from Piksi for the first time, the data will be put into a `gps_time_t` that at first was uninitialized but is now initialized, so `is_set` would transition from false to true.

If for some reason we subtract two gps times, and get a "negative" result, the negative result will have is_set = false to indicate that it's invalid.
